stl/inc/xutility

```diff
@@ -2668,9 +2744,223 @@ namespace ranges {
     // CONCEPT ranges::contiguous_range
     template <class _Rng>
     concept contiguous_range = range<_Rng> && contiguous_iterator<iterator_t<_Rng>> && requires(_Rng& __r) {
-        { data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
+        { _RANGES data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
     };
     // clang-format on
+
+    // CLASS ranges::_Not_quite_object
+    class _Not_quite_object {
+    public:
+        // Some overload sets in the library have the property that their constituent function templates are not visible
+        // to argument-dependent name lookup (ADL) and that they inhibit ADL when found via unqualified name lookup.
+        // This property allows these overload sets to be implemented as function objects. We derive such function
+        // objects from this type to remove some typical object-ish behaviors which helps users avoid depending on their
+        // non-specified object-ness.
+
+        struct _Construct_tag {};
+
+        _Not_quite_object() = delete;
+
+        constexpr explicit _Not_quite_object(_Construct_tag) noexcept {}
+
+        _Not_quite_object(const _Not_quite_object&) = delete;
+        _Not_quite_object& operator=(const _Not_quite_object&) = delete;
+
+        void operator&() const = delete;
+
+    protected:
+        ~_Not_quite_object() = default;
+    };
+
+    // CLASS ranges::_Advance_fn
+    class _Advance_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_or_output_iterator _It>
+        constexpr void operator()(_It& _Where, iter_difference_t<_It> _Off) const {
+            if constexpr (random_access_iterator<_It>) {
+                _Where += _Off;
+            } else {
+                if constexpr (!bidirectional_iterator<_It>) {
+                    _STL_ASSERT(_Off >= 0, "negative advance of non-bidirectional iterator");
+                }
+
+                auto&& _UWhere              = _Get_unwrapped_n(_STD move(_Where), _Off);
+                constexpr bool _Need_rewrap = !is_reference_v<decltype(_Get_unwrapped_n(_STD move(_Where), _Off))>;
+
+                if constexpr (bidirectional_iterator<_It>) {
+                    for (; _Off < 0; ++_Off) {
+                        --_UWhere;
+                    }
+                }

```

Comment:
Are you missing the 
```cpp
} else {
    _STL_ASSERT(_Off >= 0, "negative advance of non-bidirectional iterator");
}
``` 
here?

Reply:
It's not missing, it's ten lines up before the `_Get_unwrapped_n`. All of the `std::advance`/`ranges::advance` code paths that unwrap perform this `_STL_ASSERT` before unwrapping in hopes that doing so will make it easier to debug problems than asserting *after* the iterator has been unwrapped.

Reply:
It's not missing, it's ten lines up before the `_Get_unwrapped_n`. All of the `std::advance`/`ranges::advance` code paths that unwrap perform this `_STL_ASSERT` before unwrapping in hopes that doing so will make it easier to debug problems than asserting *after* the iterator has been unwrapped.